### ISSUE TITLE
Is it possible to get hypothesis pytest plugin to catch Unsatisfiable exception?

### ISSUE THREAD

**Issue by mulkieran** [HypothesisWorks/hypothesis-python#332]
The idea would be for it to catch the exception and somehow communicate to pytest that it should label the tests as skipped because of no test subjects rather than failed because hypothesis refuted.

**Comment by mulkieran** [comment]
I have a problem where I synthesize my examples from an environment. In a rich environment I may get many examples, in a poor one I may get none at all. So, fixing the test so that it will always get at least some examples is not possible. So, I do something like:

```
    _devices = [d for d in _DEVICES if d.parent]
    @pytest.mark.skipif(len(_devices) == 0, reason='no device with a parent')
    @given(strategies.sampled_from(_devices))
    @settings(max_examples=5, min_satisfying_examples=1)
    def ....
```
I need to make a list of valid examples, use pytest to skip if the list is empty, then construct a strategy by sampling from the list. For various reasons, I would prefer to be able to just directly make a devices strategy. But, if I do that, and the strategy yields no examples, then the test will fail with a big hypothesis stack trace, which is the wrong behavior for my purposes, which would be to just mark the test as skipped due to lack of data.

**Comment by mulkieran** [comment]
This ```is_empty``` function is a new one on me.

**Comment by DRMacIver** [comment]
Oops. Sorry I never replied to this.

I do think in general skipping unsatisfiable tests is probably the wrong thing to do. Adding a `@pytest.mark.xfail(reason=Unsatisfiable)` to tests where this is expected is probably a better approach.

> This is_empty function is a new one on me.

It's not part of the public API, but it got added [about a year ago](https://github.com/HypothesisWorks/hypothesis-python/commit/dcde0754e277bc62c6ba0727a2d202c53e3d4916). I actually thought I added in response to a request from you! But it might just be that I had some of our past discussions about sampled\_from in mind.

One caution is that it's not exactly reliable and exists more as an optimisation. A lot of things will show up as non-empty but still be empty. e.g. integers().flatmap(lambda x: nothing()) will show up as non-empty.
